Added documentation for RUM APM correlation (#9197)
DataDog/documentation: content/en/tracing/connect_rum_and_traces/_index.md

---
title: Connect RUM and Traces
kind: documentation
further_reading:
  - link: "https://www.datadoghq.com/blog/real-user-monitoring-with-datadog/"
    tag: "Blog"
    text: "Real User Monitoring"
  - link: "/tracing/"
    tag: "Documentation"
    text: "APM and Distributed Tracing"
---

{{< img src="tracing/connect_rum_and_traces/rum_trace_tab.png" alt="RUM and Traces"  style="width:100%;">}}

# Overview

The APM integration with Real User Monitoring allows you to link requests from your web and mobile applications to their corresponding backend traces, while providing the RUM session context.

> Having frontend-related specifics (thanks to your RUM session) as well as backend, infrastructure, and log information (thanks to your trace id injection) allows you to pinpoint issues anywhere in your stack in unprecedented speed with a full understanding of what your users are experiencing.

Reviewer: ```suggestion
Use frontend data from RUM, and backend, infrastructure, and log information from trace ID injection to quickly pinpoint issues anywhere in your stack and fully understand what your users are experiencing.
```

<div class="alert alert-warning">
This feature is currently available for private beta. <a href="https://docs.datadoghq.com/help/">Reach out to support</a> to turn on this feature for your account.
</div>

# Usage
## Prerequisites

> -   Your services targeted by your RUM applications are [traced on the APM side][1].

Reviewer: ```suggestion
-   Set up [APM tracing][1] on the services targeted by your RUM applications.
```

-   Your services use an HTTP server.

> -   Your HTTP servers are using a library that supports distributed tracing.

Reviewer: do we have a list of these that we can link to? I am not really sure what you mean here

Author: Should we link to the section underneath?

Reviewer: yes, perfect

## RUM set up
{{< tabs >}}
{{% tab "Browser RUM" %}}

> 1.  Set up [Browser Real User Monitoring][1]

Reviewer: ```suggestion
1.  Set up [Browser Real User Monitoring][1].
```

2. Initialize the RUM SDK. Configure the `allowedTracingOrigins` initialization parameter with the list of internal (first party) origins called by your browser application.

```javascript
import { datadogRum } from '@datadog/browser-rum'
 
datadogRum.init({
    applicationId: '<DATADOG_APPLICATION_ID>',
    clientToken: '<DATADOG_CLIENT_TOKEN>',
    ...otherConfig,
    allowedTracingOrigins: ["<https://api.example.com>", /https:\/\/.*\.my-api-domain\.com/]
})
```

**Note**: `allowedTracingOrigins` accepts Javascript String and RegExp.

[1]: /real_user_monitoring/browser/
{{% /tab %}}
{{% tab "Android RUM" %}}

> 1.  Set up [Android Real User Monitoring][1]

Reviewer: ```suggestion
1.  Set up [Android Real User Monitoring][1].
```

> 2.  Configure the OkHttpClient interceptor with the list of internal (first party) origins called by your Android application.

Reviewer: ```suggestion
2.  Configure the `OkHttpClient` interceptor with the list of internal (first party) origins called by your Android application.
```

```java
val tracedHosts =  listOf("example.com", "example.eu")

val okHttpClient = OkHttpClient.Builder()
    .addInterceptor(DatadogInterceptor(tracedHosts))
    .addNetworkInterceptor(TracingInterceptor(tracedHosts))
    .eventListenerFactory(DatadogEventListener.Factory())
    .build()
```

> **Note**: By default, all subdomains of listed hosts will be traced. For instance, adding example.com will also enable the tracing for api.example.com and foo.example.com

Reviewer: ```suggestion
**Note**: By default, all subdomains of listed hosts are traced. For instance, if you add `example.com`, you also enable the tracing for `api.example.com` and `foo.example.com`.
```

[1]: /real_user_monitoring/android/
{{% /tab %}}
{{% tab "iOS RUM" %}}

> 1.  Set up [iOS Real User Monitoring][1]

Reviewer: ```suggestion
1.  Set up [iOS Real User Monitoring][1].
```

    
2.  Set the `firstPartyHosts` initialization parameter with the list of internal (first party) origins called by your iOS application.
```swift
Datadog.initialize(
appContext: .init(),
configuration: Datadog.Configuration
    .builderUsing(rumApplicationID: "<rum_app_id>", clientToken: "<client_token>", environment: "<env_name>")
    .set(firstPartyHosts: ["example.com", "api.yourdomain.com"])
    .build()
)
```

3.  Initialize URLSession as stated in the [set up documentation][1]:
```swift
let session =  URLSession(
    configuration: ...,
    delegate: DDURLSessionDelegate(),
    delegateQueue: ...
)
```

> **Note**: By default, all subdomains of listed hosts will be traced. For instance, adding example.com will also enable the tracing for api.example.com and foo.example.com

Reviewer: ```suggestion
**Note**: By default, all subdomains of listed hosts are traced. For instance, if you add `example.com`, you also enable the tracing for `api.example.com` and `foo.example.com`.
```
